contracts/modules/staking/StakingPool.sol

```diff
@@ -1240,18 +1248,77 @@ contract StakingPool is IStakingPool, ERC721 {
     return ownerOf(0);
   }
 
-  /* management */
+  /* pool management */
+
+  function setProducts(ProductParams[] memory params) external onlyManager {
+    uint[] memory productIds = new uint[](params.length);
+
+    for (uint i = 0; i < params.length; i++) {
+      if (params[i].recalculateEffectiveWeight) {
+        productIds[i] = params[i].productId;
+      }
+    }
+
+    (uint globalCapacityRatio, uint[] memory initialPriceRatios, uint[] memory capacityReductionRatios) = ICover(coverContract).getCapacityRatios(productIds);
+
+    uint _totalTargetWeight = totalTargetWeight;
+    uint _totalEffectiveWeight = totalEffectiveWeight;
+
+    for (uint i = 0; i < params.length; i++) {
+      ProductParams memory _param = params[i];
+      StakedProduct memory _product = products[_param.productId];
+
+      if (_product.nextPriceUpdateTime == 0) {
+        _product.nextPrice = initialPriceRatios[i].toUint96();
+        _product.nextPriceUpdateTime = uint32(block.timestamp);
+      }
 
-  function addProducts(ProductParams[] memory params) external onlyManager {
-    totalSupply = totalSupply;  // To silence view fn warning. Remove once implemented
-    params;
+      if (_param.setPrice) {
+        require(_param.targetPrice <= TARGET_PRICE_DENOMINATOR, "StakingPool: Target price too high");
+        _product.targetPrice = _param.targetPrice;
+      }
+
+      // Must recalculate effectiveWeight if target weight is changed
+      if (_param.setTargetWeight && _param.recalculateEffectiveWeight) {
+        require(_param.targetWeight <= WEIGHT_DENOMINATOR, "StakingPool: Cannot set weight beyond 1");
+
+        uint8 previousEffectiveWeight = _product.lastEffectiveWeight;
+        _product.lastEffectiveWeight = _getEffectiveWeight(
+          _param.productId,
+          _param.targetWeight,
+          globalCapacityRatio,
+          capacityReductionRatios[i]
+        );
+        _totalEffectiveWeight = _totalEffectiveWeight - previousEffectiveWeight + _product.lastEffectiveWeight;
+        _totalTargetWeight = _totalTargetWeight - _product.targetWeight + _param.targetWeight;
+        _product.targetWeight = _param.targetWeight;
+      }
+      products[_param.productId] = _product;
+    }
+
+    require(_totalEffectiveWeight <= MAX_TOTAL_WEIGHT, "StakingPool: Total max effective weight exceeded");
```

Comment:
We have an edge case that might affect us in a bad way. If the effective weight is close to or at max and there's a burn - the effective weight will shoot up. If the manager wants to just lower the weight on anything but he's unable to lower below max total weight - he'll never be able to do that. Might need a special treatment to allow reducing the weights without this to revert.